src/v/cluster/partition_leaders_table.h

```diff
@@ -70,13 +92,16 @@ class partition_leaders_table {
       model::term_id term) {
         { f(tp_ns, pid, leader, term) } -> std::same_as<void>;
     }
-    void for_each_leader(Func&& f) const {
+    ss::future<> for_each_leader(Func&& f) const {
+        auto version_snapshot = _version;
         for (auto& [tp_ns, partition_leaders] : _topic_leaders) {
             for (auto& [p_id, leader_info] : partition_leaders) {
                 f(tp_ns,
                   p_id,
                   leader_info.current_leader,
                   leader_info.update_term);
+                co_await ss::coroutine::maybe_yield();
+                throw_if_modified(version_snapshot);
```

Comment:
Not sure if it's required to get always get a consistent snapshot of information, but if consistency can be relaxed, I've been using a trick to iterate over large sorted maps and preventing reactor stalls: f105ac70b89418721e7115591bc9da4bf165b702


        auto it = map.begin();
        while (it != map.end()) {
            do_iterator_action(it);
            ++it;
            if (ss::need_preempt() && it != map.end()) {
                // The iterator could have be invalidated if there was a write
                // during the yield. We'll use the ordered nature of the btree
                // to support resuming the iterator after the suspension point.
                auto checkpoint = it->first;
                co_await ss::yield();
                it = map.lower_bound(checkpoint);
            }
        }


Maybe a little annoying here with the double map, but still something to consider.

Reply:
nice little trick :)

Reply:
This is just an alternative to throwing on mutation, clearly both what is in this PR and what I have here prevent stalls.

Reply:
in this case we do want to have accurate information as leadership information propagation should be fast and reliable. It may influence partition availability as seen by the clients.